File: year_2021/solutions/julia/day_11.jl
include(joinpath(@__DIR__(), "utils", "files.jl"))


function count_flashes(grid::Matrix{Int}, steps::Int)::Int
    previous_grid = copy(grid)
    flashes = 0
    for _ in 1:steps
        step_flashes, grid_after_step = step(previous_grid)
        flashes += step_flashes
        previous_grid = grid_after_step
    end
    return flashes
end


function first_synchronized_flash(grid::Matrix{Int})::Int
    grid_area = length(grid)
    previous_grid = copy(grid)
    _step = 0
    while true
        _step += 1
        step_flashes, grid_after_step = step(previous_grid)
        if step_flashes == grid_area
            return _step
        end
        previous_grid = grid_after_step
    end
end


function step(grid::Matrix{Int})::Tuple{Int, Matrix{Int}}
    new_grid = copy(grid)
    flashed = Set{CartesianIndex}()
    new_grid = charge(grid)
    octopuses_to_flash = locate_octopuses_to_flash(new_grid, flashed)
    while length(octopuses_to_flash) > 0
        for octopus in octopuses_to_flash
            new_grid = propagate_around(octopus, new_grid)
        end
        flashed = union(flashed, octopuses_to_flash)
        octopuses_to_flash = locate_octopuses_to_flash(new_grid, flashed)
    end
    discharge!(new_grid)
    return length(flashed), new_grid
end


function locate_octopuses_to_flash(grid::Matrix{Int}, already_flashed::Set{CartesianIndex})::Set{CartesianIndex}
    octopuses_to_flash = Set()
    for idx in CartesianIndices(grid)
        if (grid[idx] > 9) & !(idx in already_flashed)
            push!(octopuses_to_flash, idx)
        end
    end
    return octopuses_to_flash
end


function propagate_around(octopus::CartesianIndex, grid::Matrix{Int})::Matrix{Int}
    new_grid = copy(grid)
    for neighbor in map(offset -> octopus + offset, CartesianIndices((-1:1, -1:1)))
        if (neighbor != octopus) & checkbounds(Bool, new_grid, neighbor)
            new_grid[neighbor] = new_grid[neighbor] + 1
        end
    end
    return new_grid
end


charge(grid::Matrix{Int})::Matrix{Int} = grid .+ 1

discharge!(grid::Matrix{Int}) = (grid[grid.>9] .= 0)


function main()
    input_file_path::String = joinpath(INPUTS_FOLDER, splitext(basename(@__FILE__))[1], "input.txt")
    octopus_grid::Matrix{Int} = read_matrix(input_file_path)

    # Part 1
    part_1_result::Int = count_flashes(octopus_grid, 100)
    @assert part_1_result == 1721
    println("Part 1 result : ", part_1_result)

    # Part 2
    part_2_result::Int = first_synchronized_flash(octopus_grid)
    @assert part_2_result == 298
    println("Part 2 result : ", part_2_result)
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: year_2021/solutions/python/day_11.py
from itertools import product
from pathlib import Path
from typing import List, Tuple, Set

from year_2021.solutions.python.utils.files import INPUTS_FOLDER, read_int_grid


def count_flashes(grid: List[List[int]], steps: int) -> int:
    previous_grid = grid.copy()
    flashes = 0
    for i in range(steps):
        step_flashes, grid_after_step = step(previous_grid)
        flashes += step_flashes
        previous_grid = grid_after_step
    return flashes


def first_synchronized_flash(grid: List[List[int]]) -> int:
    grid_area = len(grid) * len(grid[0])
    previous_grid = grid.copy()
    _step = 0
    while True:
        _step += 1
        step_flashes, grid_after_step = step(previous_grid)
        if step_flashes == grid_area:
            return _step
        previous_grid = grid_after_step


def step(grid: List[List[int]]) -> Tuple[int, List[List[int]]]:
    new_grid = grid.copy()
    flashed = set()
    new_grid = charge(new_grid)
    octopuses_to_flash = locate_octopuses_to_flash(new_grid, flashed)
    while len(octopuses_to_flash) > 0:
        for octopus in octopuses_to_flash:
            new_grid = propagate_around(octopus, new_grid)
        flashed = flashed.union(octopuses_to_flash)
        octopuses_to_flash = locate_octopuses_to_flash(new_grid, flashed)
    return len(flashed), discharge(new_grid)


def locate_octopuses_to_flash(grid: List[List[int]], already_flashed: Set[Tuple[int, int]]) -> Set[Tuple[int, int]]:
    octopuses_to_flash = set()
    for i, row in enumerate(grid):
        for j, c in enumerate(row):
            if c > 9 and (i, j) not in already_flashed:
                octopuses_to_flash.add((i, j))
    return octopuses_to_flash


def propagate_around(octopus: Tuple[int, int], grid: List[List[int]]) -> List[List[int]]:
    new_grid = grid.copy()
    offsets = [o for o in product(range(-1, 2), range(-1, 2)) if o != (0, 0)]
    for o in offsets:
        neighbor_i = octopus[0] + o[0]
        neighbor_j = octopus[1] + o[1]
        if is_in_bounds((neighbor_i, neighbor_j), grid):
            new_grid[neighbor_i][neighbor_j] = new_grid[neighbor_i][neighbor_j] + 1
    return new_grid


def is_in_bounds(neighbor: Tuple[int, int], grid: List[List[int]]) -> bool:
    rows = len(grid)
    cols = len(grid[0])
    return 0 <= neighbor[0] < rows and 0 <= neighbor[1] < cols


def charge(grid: List[List[int]]) -> List[List[int]]:
    return [[c + 1 for c in row] for row in grid]


def discharge(grid: List[List[int]]) -> List[List[int]]:
    return [[(0 if c > 9 else c) for c in row] for row in grid]


if __name__ == '__main__':
    input_file_path = INPUTS_FOLDER / Path(__file__).with_suffix('').name / 'input.txt'
    octopus_grid: List[List[int]] = read_int_grid(input_file_path=input_file_path)

    # Part 1
    part_1_result: int = count_flashes(grid=octopus_grid, steps=100)
    assert part_1_result == 1721
    print('Part 1 result :', part_1_result)

    # Part 2
    part_2_result: int = first_synchronized_flash(grid=octopus_grid)
    assert part_2_result == 298
    print('Part 2 result :', part_2_result)
